TEP-0124 implement opentelemetry and jaeger tracing

kmjayadeep opened the pull request:
Adds opentelemetry instrumentation code to pipelinerun and taskrun reconcilers. Also made required changes in the main.go to include jaeger as tracing backend

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Added support for tracing using Jaeger and OpenTelemetry. It can be enabled by adding the following environment variables to controller manifest
* `OTEL_EXPORTER_JAEGER_ENDPOINT` is the HTTP endpoint for sending spans directly to a collector.
* `OTEL_EXPORTER_JAEGER_USER` is the username to be sent as authentication to the collector endpoint. (optional)
* `OTEL_EXPORTER_JAEGER_PASSWORD` is the password to be sent as authentication to the collector endpoint. (optional)
```

afrittoli commented:
> Adds opentelemetry instrumentation code to pipelinerun and taskrun reconcilers. Also made required changes in the main.go to include jaeger as tracing backend
> 
> # Changes
> # Submitter Checklist
> As the author of this PR, please check off the items in this checklist:
> 
> * [x]  Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
> * [ ]  Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
> * [ ]  Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
> * [ ]  Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
>   functionality, content, code)
> * [ ]  Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
> * [ ]  Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
> * [ ]  Release notes contains the string "action required" if the change requires additional action from users switching to the new release
> 
> # Release Notes
> ```
> NONE
> ```

Please make sure you follow the PR template, describe changes clearly in detail under the "Changes" part, check the checklist as appropriate, and include release notes.

afrittoli commented on cmd/controller/main.go:
```diff
@@ -102,15 +112,72 @@ func main() {
 	}()
 
+	// initialize opentelemetry
+	tpPipelineRun, err := tracerProvider("pipeline-reconciler")
```
Please define a const for the trace provider names

afrittoli commented on cmd/controller/main.go:
```diff
@@ -102,15 +112,72 @@ func main() {
+func tracerProvider(service string) (trace.TracerProvider, error) {
+	// Create the Jaeger exporter
+	// The following env variables are used by the sdk for creating the exporter
+	// - OTEL_EXPORTER_JAEGER_ENDPOINT is the HTTP endpoint for sending spans directly to a collector.
+	// - OTEL_EXPORTER_JAEGER_USER is the username to be sent as authentication to the collector endpoint.
+	// - OTEL_EXPORTER_JAEGER_PASSWORD is the password to be sent as authentication to the collector endpoint.
```
The way to use this functionality today is to add these extra variables in the controller manifest.
We will need to document this, including an example. Additionally, we might want to have an example commented in the controller manifest so that to use the feature it would be enough to uncomment and set the values.

kmjayadeep replied:
added

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2019 The Tekton Authors
```
2022

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,83 @@
+	// Name of the tracer
+	Tracer = "PipelineRunReconciler"
+	// Name of the Annotation used for propogating SpanContext
+	SpanContextAnnotation = "tekton.dev/pipelinerun-span-context"
```
We don't have a clear convention for names here, but I think the majority are in camelCase, so I think `pipelineRunSpanContext` would be good.

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,83 @@
+		pro.Inject(ctxWithTrace, propagation.MapCarrier(carrier))
+
+		if !span.SpanContext().HasTraceID() {
+			logger.Debug("disabling tracing")
```
A bit more context in the message please

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,83 @@
+		logger.Info("adding spancontext", string(marshalled))
+		an[SpanContextAnnotation] = string(marshalled)
+		pr.Annotations = an
+		span.AddEvent("Updatng TaskRun CR with SpanContext annotations")
```
Updatng -> Updating

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,83 @@
+
+		marshalled, err := json.Marshal(carrier)
+		if err != nil {
+			logger.Error("Unable to marshal spancontext, err: %s", err)
```
NIT: error messages in go must start with lowercase

afrittoli replied:
this does not seem to be solved?

afrittoli commented on pkg/reconciler/pipelinerun/pipelinerun.go:
```diff
@@ -183,6 +194,8 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
 		logger.Errorf("Failed to fetch pipeline func for pipeline %s: %w", pr.Spec.PipelineRef.Name, err)
 		pr.Status.MarkFailed(ReasonCouldntGetPipeline, "Error retrieving pipeline for pipelinerun %s/%s: %s",
 			pr.Namespace, pr.Name, err)
+		span.SetStatus(codes.Error, "Failed to fetch pipelinefunc")
```
NIT: Perhaps "Failed to fetch the pipeline definition" - and include the error too?

kmjayadeep replied:
error is recorded in the line below using `span.RecordError(err)`

afrittoli commented on pkg/reconciler/pipelinerun/pipelinerun.go:
```diff
@@ -183,6 +194,8 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
 		pr.Status.MarkFailed(ReasonCouldntGetPipeline, "Error retrieving pipeline for pipelinerun %s/%s: %s",
 			pr.Namespace, pr.Name, err)
+		span.SetStatus(codes.Error, "Failed to fetch pipelinefunc")
+		span.RecordError(err)
```
Shall we record an error in all failure cases, or is there a specific reason to capture it only in case of failing to fetch the pipeline spec?

kmjayadeep replied:
we should record error wherever it makes sense to be able to debug issues by visualizing in jaeger. I just added one as an example in this PR.

afrittoli reviewed (commented):
Thanks for this!
I didn't manage to finish the review today, but I will continue tomorrow. 
Feel free to start working on the feedback if you'd like.

This PR will have to include documentation about the new feature, how to configure it and how to use it.

kmjayadeep committed:
TEP-0124 implement opentelemetry and jaeger tracing

Adds opentelemetry instrumentation code to pipelinerun and taskrun
reconcilers. Also made required changes in the main.go to include jaeger
as tracing backend

afrittoli commented on docs/developers/tracing.md:
````diff
@@ -0,0 +1,35 @@
+Jaeger should be installed and accessible from the cluster. The easiest
+way to set it up is using helm as below
+
+the following command installs Jaeger in `jaeger` namespace
+
+```
+helm repo add jaegertracing https://jaegertracing.github.io/helm-charts
+helm upgrade -i jaeger jaegertracing/jaeger -n jaeger --create-namespace
+```
````
NIT: This is really helpful, perhaps we can include a link to jaeger's own docs too

afrittoli commented on docs/developers/tracing.md:
```diff
@@ -0,0 +1,35 @@
+## Enabling tracing
+
+Tekton pipelines controller expects the following environment variables
+to be able to connect to jaeger
```
NIT missing ":"

afrittoli commented on docs/developers/tracing.md:
```diff
@@ -0,0 +1,35 @@
+
+* `OTEL_EXPORTER_JAEGER_ENDPOINT` is the HTTP endpoint for sending spans directly to a collector.
+*	`OTEL_EXPORTER_JAEGER_USER` is the username to be sent as authentication to the collector endpoint.
+*	`OTEL_EXPORTER_JAEGER_PASSWORD` is the password to be sent as authentication to the collector endpoint.
```
NIT: indentation is off here

afrittoli commented on pkg/reconciler/taskrun/tracing.go:
```diff
@@ -0,0 +1,82 @@
+	// if spanContext is not present in the CR, create new parent trace and
+	// add the marshalled spancontext to the annotations
+	if _, e := an[SpanContextAnnotation]; !e {
+		ctxWithTrace, span := tracerProvider.Tracer(Tracer).Start(ctx, "TaskRun")
```
NIT: For `PipelineRun` this is called `"PipelineRun:Reconciler"` - we should keep name consistent, and possibly defined in a const

afrittoli commented on pkg/reconciler/taskrun/tracing.go:
```diff
@@ -0,0 +1,82 @@
+
+		pro.Inject(ctxWithTrace, propagation.MapCarrier(carrier))
+
+		logger.Info("got carrier", carrier)
```
NIT: the message in this log line does not seem very helpful - also, it looks more like a debug level one?

afrittoli commented on pkg/reconciler/taskrun/tracing.go:
```diff
@@ -0,0 +1,82 @@
+			logger.Error("unable to marshal spancontext, err: %s", err)
+			return ctx
+		}
+		logger.Info("adding spancontext", "ctx", string(marshalled))
```
NIT: this feels more like a debug kind of message. 

Did you really mean to log the string "ctx"? is this meant to be the `ctx` variable?
Do you need to invoke `string()` on `marshalled`? I think `logger.Info` accepts `interface`

afrittoli commented on pkg/reconciler/taskrun/tracing.go:
```diff
@@ -0,0 +1,82 @@
+	if _, e := an[SpanContextAnnotation]; !e {
+		ctxWithTrace, span := tracerProvider.Tracer(Tracer).Start(ctx, "TaskRun")
+		defer span.End()
+		span.SetAttributes(attribute.String("task", tr.Name), attribute.String("namespace", tr.Namespace))
```
NIT: I think `taskRun` would be more precise than `task`. You could set the `task` attribute as well, in case one wanted to search for all traces associated to a specific task.

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,102 @@
+	if _, e := an[SpanContextAnnotation]; !e {
+		ctxWithTrace, span := tracerProvider.Tracer(Tracer).Start(ctx, "PipelineRun:Reconciler")
+		defer span.End()
+		span.SetAttributes(attribute.String("pipeline", pr.Name), attribute.String("namespace", pr.Namespace))
```
NIT: I think `pipelineRun` would be more precise than `pipeline`. You could set the `pipeline` attribute as well, in case one wanted to search for all traces associated to a specific task.

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,102 @@
+	logger := logging.FromContext(ctx)
+	carrier := make(map[string]string)
+
+	pro := otel.GetTextMapPropagator()
```
NIT: perhaps `propagator` instead of `pro`?

afrittoli commented on pkg/reconciler/taskrun/tracing.go:
```diff
@@ -0,0 +1,82 @@
+	logger := logging.FromContext(ctx)
+	carrier := make(map[string]string)
+
+	pro := otel.GetTextMapPropagator()
```
NIT: perhaps `propagator` instead of `pro`?

afrittoli commented on pkg/reconciler/taskrun/tracing.go:
```diff
@@ -0,0 +1,82 @@
+	carrier := make(map[string]string)
+
+	pro := otel.GetTextMapPropagator()
+	an := tr.Annotations
```
I would recommend a more intelligible name, like `trAnnotations`, or work directly on `tr.Annotations`

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,102 @@
+	carrier := make(map[string]string)
+
+	pro := otel.GetTextMapPropagator()
+	an := pr.Annotations
```
I would recommend a more intelligible name, like `prAnnotations`, or work directly on `pr.Annotations`

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,102 @@
+			logger.Error("Unable to marshal spancontext, err: %s", err)
+			return ctx
+		}
+		logger.Info("adding spancontext", marshalled)
```
Maybe `debug` instead of `info`?

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,102 @@
+		an[SpanContextAnnotation] = marshalled
+		pr.Annotations = an
+		span.AddEvent("updating TaskRun CR with SpanContext annotations")
+	}
```
NIT: I wonder if we could just return the context here, instead of un-marshalling what we just marshalled?

kmjayadeep replied:
ya makes sense

afrittoli commented on pkg/reconciler/pipelinerun/tracing.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+// Extract spanContext from the context and return it as json encoded string
+func getMarshalledSpanContext(ctx context.Context) (string, error) {
```
NIT: `getMarshalledSpanFromContext`

afrittoli reviewed (commented):
Thank you for all the updates. 
Here some more comments. 
I'm sorry I didn't manage (again) to do a full review - the comments are mostly minor things so perhaps you want to address them and start creating the PR in the pipeline repo?
I'll try to finish my review later today.

kmjayadeep commented:
Thanks a lot for the review! I have addressed the comments. I will close this PR and raise against upstream now
